File: 0983-minimum-cost-for-tickets/0983-minimum-cost-for-tickets.py
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        dp = [0] * len(days)
        dp[-1] = min(costs)
        for i in range(len(days) - 2, -1, -1):
            # 1 day pass
            dp[i] = costs[0] + dp[i + 1]

            # check 7 days pass
            idx = self.binary_search(days, i + 1, days[i] + 7) 
            dp[i] = min(dp[i], costs[1] + (dp[idx] if idx != -1 else 0))

            # check 30 days pass
            idx = self.binary_search(days, i + 1, days[i] + 30)
            dp[i] = min(dp[i], costs[2] + (dp[idx] if idx != -1 else 0))

        return dp[0]

    def binary_search(self, days, start, target):
        l, r = start, len(days) - 1
        while l < r:
            m = (l + r) // 2
            if days[m] < target:
                l = m + 1
            else:
                r = m
        if days[l] < target:
            return -1
        return l

File: 0983-minimum-cost-for-tickets/0983-minimum-cost-for-tickets.java
class Solution {
    /**
        days = [1,4,6,7,8,20], costs = [2,7,15]

        [1] 2 3 [4] 5 [6] [7] [8] 9 10.... [20]......365

        days[i]:
            buy 1day, 7day, 30day pass
        
        return min(1, 7, 30)


     */
    public int mincostTickets(int[] days, int[] costs) {
        int[] dp = new int[days.length];
        Arrays.fill(dp, -1);
        return dfs(days, 0, costs, dp);
    }

    private int dfs(int[] days, int i, int[] costs, int[] dp) {
        if (i >= days.length) 
            return 0;

        if (dp[i] != -1) return dp[i];
        
        //1 day pass
        int _1DayPass = costs[0] + dfs(days, i + 1, costs, dp);

        //7 day pass
        int d = i;
        while (d < days.length && days[d] < days[i] + 7) {
            d++;
        }
        int _7DayPass = costs[1] + dfs(days, d, costs, dp);

        //30 day pass
        d = i;
        while (d < days.length && days[d] < days[i] + 30) {
            d++;
        }
        int _30DayPass = costs[2] + dfs(days, d, costs, dp);

        return dp[i] = Math.min(_1DayPass, Math.min(_7DayPass, _30DayPass));
    }
}